patches/server/0009-Adventure.patch

```diff
@@ -524,39 +523,59 @@ index 0000000000000000000000000000000000000000..8b7b140ac0326ffccd51684e28364714
-+            this.broadcastToViewers(viewers, player, chatType, new ConstantFunction(renderedMessage));
++            this.broadcastToViewers(viewers, chatType, new ConstantFunction(renderedMessage));
 +        }
 +
 +        @Override
 +        public void sendOriginal(CraftPlayer player, Set<Audience> viewers, ChatType.Bound chatType) {
-+            this.broadcastToViewers(viewers, player, chatType, null);
++            this.broadcastToViewers(viewers, chatType, null);
 +        }
 +
-+        private void broadcastToViewers(Collection<Audience> viewers, final Player source, final ChatType.Bound chatType, final @Nullable Function<Audience, net.minecraft.network.chat.Component> msgFunction) {
-+            final Supplier<Component> fallbackSupplier = Suppliers.memoize(() -> PaperAdventure.asAdventure(msgFunction instanceof ConstantFunction constantFunction ? constantFunction.component : ChatProcessor.this.message.decoratedContent()));
-+            final Function<Audience, Component> audienceMsgFunction = !(msgFunction instanceof ConstantFunction || msgFunction == null) ? msgFunction.andThen(PaperAdventure::asAdventure) : viewer -> fallbackSupplier.get();
++        private void broadcastToViewers(Collection<Audience> viewers, final ChatType.Bound chatType, final @Nullable Function<Audience, net.minecraft.network.chat.Component> msgFunction) {
 +            for (Audience viewer : viewers) {
-+                if (viewer instanceof Player || viewer instanceof ConsoleCommandSender) {
-+                    // players and console have builtin PlayerChatMessage sending support while other audiences do not
-+                    this.sendToViewer((CommandSender) viewer, chatType, msgFunction);
++                if (acceptsNative(viewer)) {
++                    this.sendNative(viewer, chatType, msgFunction);
 +                } else {
-+                    viewer.sendMessage(source, audienceMsgFunction.apply(viewer), MessageType.CHAT);
++                    final net.minecraft.network.chat.@Nullable Component unsigned = Util.mapNullable(msgFunction, f -> f.apply(viewer));
++                    final PlayerChatMessage msg = unsigned == null ? ChatProcessor.this.message : ChatProcessor.this.message.withUnsignedContent(unsigned);
++                    viewer.sendMessage(msg.adventureView(), this.adventure(chatType));
 +                }
 +            }
 +        }
 +
-+        private void sendToViewer(final CommandSender viewer, final ChatType.Bound chatType, final @Nullable Function<Audience, net.minecraft.network.chat.Component> msgFunction) {
++        private net.kyori.adventure.chat.ChatType.Bound adventure(ChatType.Bound chatType) {
++            return net.kyori.adventure.chat.ChatType.chatType(
++                Key.key(ChatProcessor.this.server.registryAccess().registryOrThrow(Registries.CHAT_TYPE).getKey(chatType.chatType()).toString())
```

Comment:
It it worth trying to use the built-in instances of ChatType here if the chat type is one of the built-ins?